libraries/DataFlash/DataFlash_File.cpp

```diff
@@ -133,38 +128,27 @@ void DataFlash_File::Init()
         return;
     }
 #endif
-    
-    if (_writebuf != NULL) {
-        free(_writebuf);
-        _writebuf = NULL;
-    }
 
     // determine and limit file backend buffersize
-    uint8_t bufsize = _front._params.file_bufsize;
+    uint32_t bufsize = _front._params.file_bufsize;
     if (bufsize > 64) {
-        // we currently use uint16_t for the ringbuffer.  Also,
-        // PixHawk has DMA limitaitons.
-        bufsize = 64;
+        bufsize = 64; // Also, PixHawk has DMA limitaitons.
     }
-    _writebuf_size = bufsize * 1024;
+    bufsize *= 1024;
 
-    /*
-      if we can't allocate the full writebuf then try reducing it
-      until we can allocate it
-     */
-    while (_writebuf == NULL && _writebuf_size >= _writebuf_chunk) {
-        hal.console->printf("DataFlash_File: buffer size=%u\n", (unsigned)_writebuf_size);
-        _writebuf = (uint8_t *)malloc(_writebuf_size);
-        if (_writebuf == NULL) {
-            _writebuf_size /= 2;
-        }
-    }
-    if (_writebuf == NULL) {
+    // If we can't allocate the full size, try to reduce it until we can allocate it
+    do {
+        hal.console->printf("DataFlash_File: buffer size=%u\n", (unsigned)bufsize);
+        _writebuf.set_size(bufsize);
+        bufsize >>= 1;
```

Comment:
Which requires changing ::set_size as well because it doesn't return anything so you can't do the check you proposed.

I do prefer to have a version of ::set_size that returns a boolean. But if for some reason it gets decided that this method should not chance, I do prefer my current implementation (and also don't see a problem with buffsize value).